src/dbnode/persist/fs/seek_manager.go

```diff
@@ -498,6 +504,68 @@ func (m *seekerManager) updateOpenLeaseHotSwapSeekers(
 	return wg, updateOpenLeaseResult, nil
 }
 
+func (m *seekerManager) RelinquishShard(shard uint32) error {
+	byTime := m.seekersByTime(shard)
+
+	// Indicate this shard is closed so it can't be borrowed and no seekers
+	// can be opened from now on.
+	byTime.Lock()
```

Comment:
Since we're talking about holding this lock for longer I think you'll need to change this to be `acquireByTimeLockWaitGroupAware`

Reply:
I think `acquireByTimeLockWaitGroupAware` returns only a subset of seekers (from a specific blockStart) so we would need to call it multiple times when we iterate over all of them right? So after each iteration we will have to unlock again (see below).

With that in mind I though we lock first, mark as closed (so they can't be borrowed anymore), unlock again, the start the `acquireByTimeLockWaitGroupAware` loop to wait for remaining WaitGroups.

Then loop again to close.